text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,305 @@
+
+Generally, Hubs and Scopes are a very Sentry-specific concept, and can be a bit hard to explain. 
+Also, how Hub & Scope forking behaves is currently a bit tricky, and does not play well e.g. with how OpenTelemetry contexts work.
+
+This RFC aims to streamline this by merging the concepts of Hub and Scope into a singular Scope concept.
+
+It also proposes the new concepts of global & isolated scopes.
+
+# Background
+
+TODO
+
+# Proposed Solution
+
+Hubs will be removed as a concept for SDK users. 
+Instead, from a users perspective only Scopes remain, which will become the singular entity to hold context data etc.
+
+Scopes will be _similar_ to how they work today, but not entirely the same. 
+Scopes can have data (e.g. tags, user, ...) added to them the same way you can do today. 
+This RFC _does not_ aim to change any of the data that is kept on the scope and is applied to events.
+
+The following APIs will be removed/deprecated:
+
+* `getCurrentHub()`
+* `configureScope()` (instead just get the scope and set on it directly)
+* Any APIs currently on the Hub only: 
+  * `hub.pushScope()`
+  * `hub.popScope()`
+  * `hub.isOlderThan()`
+  * `hub.bindClient()`
+  * `hub.getStack()`
+  * `hub.getStackTop()`
+  * `hub.run()` (use `withScope()` instead)
+
+Instead, we will introduce some new APIs:
+
+```ts
+// get the currently active scope. replacement for `getCurrentHub().getScope()`
+export function getScope(): Scope;
+
+// get the currently active client. May return a NOOP client. Replacement for `getCurrentHub().getClient()`.
+export function getClient(): Client;
+
+// make a scope the current scope. Replacement for `makeMain(hub)`
+export function setCurrentScope(scope: Scope): void;
+
+// get the currently active global scope
+export function getGlobalScope(): Scope;
+
+// get the currently active isolation scope
+export function getIsolationScope(): Scope;
+
+// similar to `withScope`, but defines an isolation scope
+export function withIsolationScope(callback: (scope: Scope) => unknown): unknown;
+```
+
+The following APIs already exist but will behave differently:
+
+* `withScope()` will still work, but it will actually fork an execution context. So this will roughly do the same as doing `hub.run()` today in languages that have that, which forks an execution context. 
+
+APIs that are currently on the hub should instead be called directly on the scope (e.g. `scope.captureException()` etc.), or via a global method (e.g. `Sentry.captureException()`).
+
+The current scope may be kept similar to how we currently keep the current hub, but this is SDK specific and not part of this RFC.
+
+## Clients
+
+Instead of a client being optional, there will now _always_ be a client. It may be a Noop Client that does nothing, if `init()` has not been called yet.
+
+A scope has a reference to a client. By default it will reference a noop client. You can bind a client to a scope via `scope.setClient()`. 
+The client is inherited by forked scopes.
```

Comment:
I think `scope.setClient()` is potentially dangerous to have due to its mutable nature. What use-cases do we need it for? If possible I would avoid it at all cost.

IMO much better would be to do `client.withScope()` for any cases where `scope.setClient()` would currently be applicable.

Reply:
I think `client.withScope` is confusing and raises questions. The client does not own scopes, so why would it make up one to bind itself? Logically the client now is a property of the scope like any other thing. Rebinding it on the scope is fine. The client on the scope needs to be mutable anyways to make rebinding possible, even if it's just internal. So we don't win anything by not exposing this.

Reply:
I am sus but let's go for `scope.setClient()` then. 😄 My FP brain doesn't like this API - it's too imperative.